### PR TITLE
Added API to provide mutiple regex patterns for enumerating files.

### PR DESCRIPTION
**Pull request by harshjain2**
Added API to provide mutiple regex patterns for enumerating files.
This should give perf benefit as Directory.EnumerateFiles is called only once and files names are matched using multiple regex patterns. 

It is bound to give an improvement of around 5-8%

### CODE REVIEW

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 1]
```diff
@@ -175,15 +175,17 @@ private void AddExtensionAssemblies(string runSettings)
...
                         continue;
                     }
 
-                    var extensionAssemblies = new List<string>(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.TestAdapterRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.TestLoggerRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.RunTimeRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.SettingsProviderRegexPattern, SearchOption.AllDirectories));
+                    var patterns = new string[] { TestPlatformConstants.TestAdapterRegexPattern, TestPlatformConstants.TestLoggerRegexPattern, TestPlatformConstants.RunTimeRegexPattern, TestPlatformConstants.SettingsProviderRegexPattern };
 
+                    var extensionAssemblies = new List<string>(this.fileHelper.EnumerateFiles(adapterPath, patterns, SearchOption.AllDirectories));
```
Comment (codito) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: Not sure if the responsibility to combine and create a regex should be with EnumerateFiles.

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 2]
```diff
@@ -175,15 +175,17 @@ private void AddExtensionAssemblies(string runSettings)
...
+                        }
+
                         continue;
                     }
 
-                    var extensionAssemblies = new List<string>(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.TestAdapterRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.TestLoggerRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.RunTimeRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.SettingsProviderRegexPattern, SearchOption.AllDirectories));
+                    var patterns = new string[] { TestPlatformConstants.TestAdapterRegexPattern, TestPlatformConstants.TestLoggerRegexPattern, TestPlatformConstants.RunTimeRegexPattern, TestPlatformConstants.SettingsProviderRegexPattern };
```
Comment (codito) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: If all these patterns are known why not Regex.Compile it directly in this class. We will save some time during runtime?

**Review by codito — CHANGES_REQUESTED** [review]
Suggest few changes.

**Comment by AbhitejJohn** [comment]
quick question: How much of a perf benefit did this net us? Just curious.

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 3]
```diff
@@ -175,15 +176,17 @@ private void AddExtensionAssemblies(string runSettings)
...
                         continue;
                     }
 
-                    var extensionAssemblies = new List<string>(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.TestAdapterRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.TestLoggerRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.RunTimeRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.SettingsProviderRegexPattern, SearchOption.AllDirectories));
-
+                    var patterns = new string[] { TestPlatformConstants.TestAdapterRegexPattern, TestPlatformConstants.TestLoggerRegexPattern, TestPlatformConstants.RunTimeRegexPattern, TestPlatformConstants.SettingsProviderRegexPattern };
+                    var regex = new Regex(string.Join("|", patterns), RegexOptions.Compiled | RegexOptions.IgnoreCase);
```
Comment (codito) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: Can this regex be a constant?

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 4]
```diff
@@ -175,15 +176,17 @@ private void AddExtensionAssemblies(string runSettings)
...
+
                         continue;
                     }
 
-                    var extensionAssemblies = new List<string>(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.TestAdapterRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.TestLoggerRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.RunTimeRegexPattern, SearchOption.AllDirectories));
-                    extensionAssemblies.AddRange(this.fileHelper.EnumerateFiles(adapterPath, TestPlatformConstants.SettingsProviderRegexPattern, SearchOption.AllDirectories));
-
+                    var patterns = new string[] { TestPlatformConstants.TestAdapterRegexPattern, TestPlatformConstants.TestLoggerRegexPattern, TestPlatformConstants.RunTimeRegexPattern, TestPlatformConstants.SettingsProviderRegexPattern };
```
Comment (codito) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: nit: runtime is one word

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManager.cs [hunk 5]
```diff
@@ -184,8 +184,7 @@ private void InitializeExtensions(IEnumerable<string> sources)
 
             if (TestPluginCache.Instance.PathToExtensions != null)
             {
-                var regex = new Regex(TestPlatformConstants.TestAdapterRegexPattern, RegexOptions.IgnoreCase);
-                extensions.AddRange(TestPluginCache.Instance.PathToExtensions.Where(ext => regex.IsMatch(ext)));
+                extensions.AddRange(TestPluginCache.Instance.PathToExtensions.Where(ext => TestExecutorExtensionManager.Regex.IsMatch(ext)));
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManager.cs]: Should these be different regular expression than the discovery one?

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/DataCollectionExtensionManager.cs [hunk 6]
```diff
@@ -17,6 +18,8 @@ namespace Microsoft.VisualStudio.TestPlatform.Common.ExtensionFramework
     /// </summary>
     internal class DataCollectorExtensionManager : TestExtensionManager<DataCollector, IDataCollectorCapabilities>
     {
+        public static Regex Regex = new Regex(TestPlatformConstants.DataCollectorRegexPattern, RegexOptions.Compiled | RegexOptions.IgnoreCase);
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/DataCollectionExtensionManager.cs]: nit: can be more specific name?

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs (hunk 4 above)
Comment (codito) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: Most of the patterns are compatible with `string.EndsWith`. Do we know how does string comparison compare with regex? Is there any observable startup time impact in regex compiled?

Comment (harshjain2) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: On measuring perf, it is observed that `String.EndsWith` performs better than `Regex`, therefore switched to `String.EndsWith` approach.

**Review by codito — APPROVED** [review]

**Pull request merged**